### DISCUSSION TITLE
Lost updates (randomly) using fastapi

### DISCUSSION THREAD

**Discussion by drc0** [sqlalchemy/sqlalchemy#8908] [Usage Questions]
hello,

we have a code like this inside fastapi 0.65.1, sqlalchemy 1.4.35, using postgres as db , sometimes the update in this code fails silently, we have some check (print) and it seems the passed data is correct..

```python
@controller.post("") 
def something(db: Session = Depends(get_db)): 
    try:
        callsomething(db) 
    except Exception:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="server_error",
        )

def callsomething(db): 
  res = db.execute(f"select mytable") 
  sql = "update mytable .... {res[0]} .... update mytable2..."
  db.execute(sql)
  db.commit() 
  print(sql) # seems ok
  return True
```
where get_db is:

```python
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine) 
def get_db(): 
  db = SessionLocal() 
  try: 
    yield 
    db 
  finally: 
    db.close()
```

does that ring a bell to anyone? we also are unable to reproduce the bug with something repetible, it happens daily on thousand requests a week..

**EDIT**

noticed something else

we have an http middleware that does some logging in the db with something like this

```python
db = SessionLocal()
try:
   response = await call_next(request)
   ...
except:
    if something_based_on_current_route: 
      try:
      finally:
         db.close()
return response
```

I'm worried about that db.close not always being executed, it is only skipped on some internal routed, could that affect something?

**Comment by zzzeek** [comment]
how do you know it fails, and not that the row is deleted or updated again elsewhere?     if you want to check that the UPDATE succeeded, you'd want to run a SQL check outside of the "commit".  if the commit happened, and then a SELECT in a new transaction after that shows the row is there, then it worked.

**Reply by drc0** [reply]
@zzzeek thanks for your answer!
that record is modified only from that query, we happen to notice it because some functionality in the app breaks. we are introducing more checks on that but not being able to reproduce the issue we are in 'waiting'...

**Reply by drc0** [reply]
@zzzeek I edited my original question with some other information about a middleware

**Reply by zzzeek** [reply]
db.close() is not very critical if "db" itself is falling out of scope in any case, this would impact more how connections are cleaned up and returned to the pool but would not affect UPDATE statements being committed or not

**Reply by drc0** [reply]
two questions:
the select for testing if the update suceeded, could it happen after the db.commit where I have `print(sql) # seems ok`  or should it happen after all those `db = SessionLocal()` , do they create a postgres begin transaction even if I have autocommit=False ?

can you expand on how connections cleanup is affected if not closing db session in that middleware?

**Reply by CaselIT** [reply]
like mike says add a select after the commit with an assert of some kind to verify that update worked or not.
If there's an error there then you are sure that something funny is happening

regarding fastapi integration from the code you have shown it does not seem to influence things, since commit happens right after the execute, so the connection should be the same as the one that executed the update

**Reply by drc0** [reply]
from the postgres logs I see many  `[11502]:LOG: could not receive data from client: Connection reset by peer` could this be a connection / pooling issue?

**Reply by zzzeek** [reply]
that means there's a dropped connection, which probably happens when someone presses the "STOP" button on their browser and interrupts the web request in progress.    The means by which SQLAlchemy closes the connection here is driver dependent.  when using asyncpg , it uses connection.terminate().

There's nothing "wrong" with that message per se, it depends upon what made it happen if it's a problem or not.
